### ISSUE TITLE
Using parse/format with AutoCompleteArrayInput and ReferenceArrayInput

### ISSUE THREAD

**Issue by estephan** [marmelab/react-admin#4454]
I'm using a `ReferenceArrayInput` with `AutocompleteArrayInput` as its child and using `parse` and `format` on the `ReferenceArrayInput`.
If the field already has a value, I cannot type into the autocomplete input for it to filter my results.
It seems to just take my first character and then just clear the filter value, meaning I cannot search on a string with more than one character.

**Steps to reproduce:**

Using [this](https://codesandbox.io/s/billowing-fire-zqytt) code sandbox:
1. Head to the post list page and to the edit page of the "Fusce massa lorem, pulvinar a posuere ut, accumsan ac nisi" post.
2. Go to the tags field on the "Miscellaneous" tab.
3. Try search for the "Rock" tag by typing "Roc" into the `AutocompleteArrayInput`.
4. Notice how the suggestions change to support your "R" filter value, but then quickly chnage back as if the filter value is cleared.

https://codesandbox.io/s/billowing-fire-zqytt

**Other information:**

There exists [this other issue](https://github.com/marmelab/react-admin/issues/4125) that happens to use the exact same components with the `parse/format` props. While that issue has been resolved, I feel that I might be missing something otherwise @Bnaya would have noticed this exact issue I'm reporting here.

I feel this issue could have to do with [this line](https://github.com/marmelab/react-admin/blob/b6a403af1f0d54c3457dd43a518e7a2ff81df338/packages/ra-ui-materialui/src/input/AutocompleteArrayInput.tsx#L217) which clears the filter when `input.value` changes. And considering `input.value` is an array and it changes in my example `format` function (`format={v => v && v.map(x => x._id)}`) as I always return a new array, I think `handleFilterChange('');` always gets called. But I haven't looked too much into this. Just a guess.

**Environment**

* React-admin version: 3.2.3
* React version: 16.13.0
* Browser: Chrome Version 79.0.3945.130

**Comment by ttlong3103** [comment]
Same issue here

**Comment by estephan** [comment]
Not sure I understood exactly what you meant.

I removed the `format` prop from `ReferenceArrayInput` and added it to the child `AutocompleteArrayInput`.

As shown in the "Tags" input in the attached image, the input can't render the tag name.

![Screen Shot 2020-02-28 at 2 12 13 pm](https://user-images.githubusercontent.com/3270818/75507272-bd5cdb80-5a34-11ea-9b6f-42a89f839877.png)

**Comment by ttlong3103** [comment]
Since `v.map()` will return a new reference which causes re-render and lose the filter (or I think so), you should return the same reference as the following:
```
format={v => {
            for (let i = 0; i < v.length; i++) {
              if (typeof v[i] == 'object') {
                v[i] = v[i].label.id;
              }
            }
            return v; // return same reference 
          }}
```

**Comment by estephan** [comment]
When I saw [this line](https://github.com/marmelab/react-admin/blob/b6a403af1f0d54c3457dd43a518e7a2ff81df338/packages/ra-ui-materialui/src/input/AutocompleteArrayInput.tsx#L217) I tried to change just the values of the elements in the array and not the reference of the array but that meant by the end of it, the value in the form was never the "parsed" value and always the "formatted" one.

If I simply load a  `ReferenceArrayInput` and select the first option, this is the order that `parse` and `format` get called. And considering `format` changes the array (which I believe is `input.value`, the forms value changes and is no longer the parsed one.

![image](https://user-images.githubusercontent.com/3270818/75510257-770c7a00-5a3e-11ea-9ce5-3988afdc2bf4.png)

**Comment by ttlong3103** [comment]
Did you try putting `parse()` in `AutocompleteArrayInput` instead of `ReferenceArrayInput`?
https://codesandbox.io/s/nifty-galois-cwr3d

**Comment by estephan** [comment]
@ttlong3103 I did try it. No luck.

**Comment by Kmaschta** [comment]
@estephan Your intuition was good, it was an issue with the dependency array of the filter reset effect.
I fixed it with #4653, it should be good for the next release :)
